File: sdk/go/tunr.go
// Package tunr is the official Go SDK for tunr.
//
// Use it to programmatically create tunnels from CI/CD pipelines,
// test automation, or custom tooling.
//
// Usage:
//
//	client, err := tunr.NewClient()
//	tunnel, err := client.Share(ctx, 3000, tunr.Options{})
//	fmt.Println(tunnel.PublicURL) // https://abc123.tunr.sh
//	defer tunnel.Close()
package tunr

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os/exec"
	"time"
)

// Client talks to the tunr API. Works via both REST and CLI under the hood.
type Client struct {
	token      string
	apiBase    string
	httpClient *http.Client
}

// Option configures the client
type Option func(*Client)

// WithToken sets the auth token
func WithToken(token string) Option {
	return func(c *Client) { c.token = token }
}

// WithAPIBase sets a custom relay address (for self-hosted deployments)
func WithAPIBase(base string) Option {
	return func(c *Client) { c.apiBase = base }
}

// NewClient creates a tunr client
func NewClient(opts ...Option) (*Client, error) {
	c := &Client{
		apiBase: "https://relay.tunr.sh",
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
	for _, opt := range opts {
		opt(c)
	}
	return c, nil
}

// Tunnel represents an active tunnel
type Tunnel struct {
	ID        string
	PublicURL string
	LocalPort int
	StartedAt time.Time

	done chan struct{}
	cmd  *exec.Cmd
}

// TunnelOptions configures tunnel creation
type TunnelOptions struct {
	Subdomain string
	NoInspect bool
}

// Share exposes a local port as a public HTTPS URL.
//
//	tunnel, err := client.Share(ctx, 3000, TunnelOptions{})
//	// tunnel.PublicURL = "https://abc123.tunr.sh"
//	defer tunnel.Close()
func (c *Client) Share(ctx context.Context, port int, opts TunnelOptions) (*Tunnel, error) {
	if port < 1024 || port > 65535 {
		return nil, fmt.Errorf("invalid port: %d", port)
	}

	args := []string{"share", "--port", fmt.Sprintf("%d", port), "--no-open"}
	if opts.Subdomain != "" {
		args = append(args, "--subdomain", opts.Subdomain)
	}

	cmd := exec.CommandContext(ctx, "tunr", args...)

	urlCh := make(chan string, 1)
	errCh := make(chan error, 1)

	stdout, _ := cmd.StdoutPipe()
	stderr, _ := cmd.StderrPipe()

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start tunr: %w", err)
	}

	readOutput := func(r io.Reader) {
		buf := make([]byte, 1024)
		full := ""
		for {
			n, err := r.Read(buf)
			if n > 0 {
				full += string(buf[:n])
				for _, line := range splitLines(full) {
					if url := extractURL(line); url != "" {
						select {
						case urlCh <- url:
						default:
						}
						return
					}
				}
			}
			if err != nil {
				return
			}
		}
	}

	go readOutput(stdout)
	go readOutput(stderr)

	var publicURL string
	select {
	case publicURL = <-urlCh:
	case err := <-errCh:
		return nil, err
	case <-time.After(15 * time.Second):
		cmd.Process.Kill()
		return nil, fmt.Errorf("tunnel failed to start within 15 seconds")
	case <-ctx.Done():
		cmd.Process.Kill()
		return nil, ctx.Err()
	}

	tunnel := &Tunnel{
		PublicURL: publicURL,
		LocalPort: port,
		StartedAt: time.Now(),
		done:      make(chan struct{}),
		cmd:       cmd,
	}

	go func() {
		cmd.Wait()
		close(tunnel.done)
	}()

	return tunnel, nil
}

// Close shuts down the tunnel
func (t *Tunnel) Close() error {
	if t.cmd != nil && t.cmd.Process != nil {
		return t.cmd.Process.Kill()
	}
	return nil
}

// Done returns a channel that closes when the tunnel exits
func (t *Tunnel) Done() <-chan struct{} {
	return t.done
}

// ─── Inspector API ────────────────────────────────────────────────────────────

// Request is a captured HTTP request from the inspector
type Request struct {
	ID         string            `json:"id"`
	Method     string            `json:"method"`
	Path       string            `json:"path"`
	StatusCode int               `json:"status_code"`
	DurationMs int64             `json:"duration_ms"`
	Timestamp  time.Time         `json:"timestamp"`
	ReqBody    string            `json:"req_body"`
	RespBody   string            `json:"resp_body"`
	ReqHeaders map[string]string `json:"req_headers"`
}

// Requests fetches recent HTTP requests from the inspector
func (c *Client) Requests(ctx context.Context, limit int) ([]*Request, error) {
	dashURL := "http://localhost:19842/api/v1/requests"
	if limit > 0 {
		dashURL += fmt.Sprintf("?limit=%d", limit)
	}

	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, dashURL, nil)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("could not reach inspector (is the daemon running?): %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Requests []*Request `json:"requests"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Requests, nil
}

// Replay resends a captured request to the local server
func (c *Client) Replay(ctx context.Context, requestID string, port int) error {
	url := fmt.Sprintf("http://localhost:19842/api/v1/requests/%s?action=replay&port=%d",
		requestID, port)
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, url, nil)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("replay failed (status %d)", resp.StatusCode)
	}
	return nil
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

func extractURL(line string) string {
	for _, word := range splitWords(line) {
		if len(word) > 8 &&
			(startsWith(word, "https://") || startsWith(word, "http://")) &&
			contains(word, "tunr.sh") {
			return word
		}
	}
	return ""
}

func splitLines(s string) []string {
	var lines []string
	start := 0
	for i, c := range s {
		if c == '\n' {
			lines = append(lines, s[start:i])
			start = i + 1
		}
	}
	return lines
}

func splitWords(s string) []string {
	var words []string
	start := -1
	for i, c := range s {
		if c == ' ' || c == '\t' || c == '\n' || c == '\r' {
			if start >= 0 {
				words = append(words, s[start:i])
				start = -1
			}
		} else if start < 0 {
			start = i
		}
	}
	if start >= 0 {
		words = append(words, s[start:])
	}
	return words
}

func startsWith(s, prefix string) bool {
	return len(s) >= len(prefix) && s[:len(prefix)] == prefix
}

func contains(s, sub string) bool {
	for i := 0; i <= len(s)-len(sub); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

File: sdk/js/tunr.js
/**
 * @tunr/sdk — JavaScript/TypeScript SDK
 * 
 * Use the tunr CLI programmatically from Node.js applications.
 * 
 * @example
 * import { TunrClient } from '@tunr/sdk';
 * 
 * const tunr = new TunrClient();
 * const tunnel = await tunr.share(3000);
 * console.log(tunnel.publicUrl); // https://abc123.tunr.sh
 * 
 * // For Express/Next.js test servers:
 * const url = await tunr.shareServer(app, { port: 3000 });
 */

'use strict';

const { spawn, exec } = require('child_process');
const http = require('http');
const https = require('https');

class TunrClient {
    /**
     * @param {Object} options
     * @param {string} [options.token] - Auth token (optional)
     * @param {string} [options.binary='tunr'] - Path to the tunr binary
     * @param {number} [options.dashPort=19842] - Inspector dashboard port
     */
    constructor(options = {}) {
        this.token = options.token;
        this.binary = options.binary || 'tunr';
        this.dashPort = options.dashPort || 19842;
        this._activeTunnels = [];
    }

    /**
     * Share a local port as a public URL
     * @param {number} port - Local port (e.g. 3000)
     * @param {Object} [opts] - Options
     * @param {string} [opts.subdomain] - Preferred subdomain
     * @param {number} [opts.timeout=15000] - Tunnel startup timeout (ms)
     * @returns {Promise<Tunnel>}
     */
    async share(port, opts = {}) {
        if (port < 1024 || port > 65535) {
            throw new Error(`Invalid port: ${port} (must be between 1024-65535)`);
        }

        const args = ['share', '--port', String(port), '--no-open'];
        if (opts.subdomain) args.push('--subdomain', opts.subdomain);

        return new Promise((resolve, reject) => {
            const proc = spawn(this.binary, args, {
                stdio: ['ignore', 'pipe', 'pipe'],
            });

            let resolved = false;
            const timeout = setTimeout(() => {
                if (!resolved) {
                    proc.kill();
                    reject(new Error('Tunnel failed to start within 15 seconds. Run `tunr doctor`.'));
                }
            }, opts.timeout || 15000);

            const tryExtractURL = (text) => {
                const match = text.match(/https?:\/\/[^\s]+tunr\.dev[^\s]*/);
                if (match && !resolved) {
                    resolved = true;
                    clearTimeout(timeout);
                    const tunnel = new Tunnel(match[0], port, proc);
                    this._activeTunnels.push(tunnel);
                    resolve(tunnel);
                }
            };

            proc.stdout.on('data', (d) => tryExtractURL(d.toString()));
            proc.stderr.on('data', (d) => tryExtractURL(d.toString()));

            proc.on('error', (err) => {
                clearTimeout(timeout);
                reject(new Error(`Failed to start tunr: ${err.message}. Is tunr in your PATH?`));
            });

            proc.on('exit', (code) => {
                clearTimeout(timeout);
                if (!resolved) {
                    reject(new Error(`tunr exited unexpectedly (exit ${code})`));
                }
            });
        });
    }

    /**
     * Share an HTTP server (for Express, Fastify, Next.js)
     * @param {http.Server|Object} server - http.Server or Express app
     * @param {Object} [opts]
     * @param {number} [opts.port=3000]
     * @returns {Promise<string>} Public URL
     * 
     * @example
     * const app = express();
     * const url = await tunr.shareServer(app, { port: 3000 });
     * // url = "https://abc123.tunr.sh"
     */
    async shareServer(server, opts = {}) {
        const port = opts.port || 3000;

        // If it's an Express app, start listening
        if (typeof server.listen === 'function' && !server.listening) {
            await new Promise((resolve) => server.listen(port, resolve));
        }

        const tunnel = await this.share(port, opts);
        return tunnel.publicUrl;
    }

    /**
     * Fetch recent HTTP requests
     * @param {number} [limit=10]
     * @returns {Promise<Array>}
     */
    async requests(limit = 10) {
        return new Promise((resolve, reject) => {
            const reqOpts = {
                hostname: 'localhost',
                port: this.dashPort,
                path: `/api/v1/requests?limit=${limit}`,
                method: 'GET',
            };
            const req = http.request(reqOpts, (res) => {
                let data = '';
                res.on('data', (d) => (data += d));
                res.on('end', () => {
                    try {
                        resolve(JSON.parse(data).requests || []);
                    } catch {
                        resolve([]);
                    }
                });
            });
            req.on('error', () => resolve([]));
            req.end();
        });
    }

    /**
     * Close all active tunnels
     */
    async closeAll() {
        await Promise.all(this._activeTunnels.map((t) => t.close()));
        this._activeTunnels = [];
    }
}

/**
 * An active tunnel
 */
class Tunnel {
    constructor(publicUrl, localPort, process) {
        this.publicUrl = publicUrl;
        this.localPort = localPort;
        this._process = process;
        this._closed = false;

        // Mark as closed when the process exits
        process.on('exit', () => { this._closed = true; });
    }

    /** Tunnel URL */
    get url() { return this.publicUrl; }

    /** Is the tunnel alive? */
    get isAlive() { return !this._closed; }

    /** Close the tunnel */
    async close() {
        if (this._process && !this._closed) {
            this._process.kill('SIGTERM');
            this._closed = true;
        }
    }
}

module.exports = { TunrClient, Tunnel };

// ─── Jest/Vitest Test Helper ──────────────────────────────────────────────────

/**
 * Jest/Vitest integration test helper
 * 
 * @example
 * // jest.setup.js:
 * import { withTunnel } from '@tunr/sdk/test';
 * 
 * describe('Webhook test', () => {
 *   let tunnelURL;
 *   
 *   beforeAll(async () => {
 *     tunnelURL = await withTunnel(3000, async (url) => {
 *       // Set up the Stripe/Paddle webhook endpoint
 *       await stripe.webhooks.update({ url });
 *     });
 *   });
 * });
 */
async function withTunnel(port, callback) {
    const client = new TunrClient();
    const tunnel = await client.share(port);
    try {
        await callback(tunnel.publicUrl, tunnel);
    } finally {
        await tunnel.close();
    }
}

module.exports.withTunnel = withTunnel;
